### ISSUE TITLE
Multiple issues in Agentless get started tutorial

### ISSUE THREAD

**Issue by warpr** [saltstack/salt#44688]
### TL;DR

> Probably the main thing which should be clarified in the documentation at https://docs.saltstack.com/en/getstarted/ssh/index.html is the role of the `master` and `Saltfile` files, and which options go in which file.  It is completely unclear to me why `config_dir` and `roster_file` can go in the Saltfile, but all the other paths need to be in the master file.

### Description of Issue/Question

Hello, I tried to follow the get started documentation and ran into many issues.  I thought it might be useful to relay my experience to illustrate at which steps the documentation seems unclear.

If this github issue is not focused or actionable enough, feel free to just close it.  I'm just submitting in the hope of it being useful in case someone is working on improving the documentation.

Here are the steps I followed:

1.  Let's give salt a try, google "saltstack getting started"
2.  Google directs me to https://docs.saltstack.com/en/getstarted/
3.  I just want to deploy some stuff to a single VPS, so let's try "Agentless SaltStack", that seems the least complicated option
4.  Clicking on "Agentless SaltStack" took me to https://docs.saltstack.com/en/getstarted/ssh/index.html, let's follow along:

```sh
06:52 kuno@18-volt:~/code$ mkdir salt
06:52 kuno@18-volt:~/code$ cd salt
06:53 kuno@18-volt:~/code/salt$ virtualenv-3 ve
Using base prefix '/usr'
New python executable in /home/kuno/code/salt/ve/bin/python3
Also creating executable in /home/kuno/code/salt/ve/bin/python
Installing setuptools, pip, wheel...done.
06:53 kuno@18-volt:~/code/salt$ source ve/bin/activate
(ve) 06:53 kuno@18-volt:~/code/salt$ pip install salt-ssh
Collecting salt-ssh
Collecting MarkupSafe (from salt-ssh)
Collecting msgpack-python>0.3 (from salt-ssh)
Collecting requests>=1.0.0 (from salt-ssh)
  Using cached requests-2.18.4-py2.py3-none-any.whl
Collecting tornado>=4.2.1 (from salt-ssh)
Collecting Jinja2 (from salt-ssh)
  Using cached Jinja2-2.10-py2.py3-none-any.whl
Collecting PyYAML (from salt-ssh)
Collecting urllib3<1.23,>=1.21.1 (from requests>=1.0.0->salt-ssh)
  Using cached urllib3-1.22-py2.py3-none-any.whl
Collecting idna<2.7,>=2.5 (from requests>=1.0.0->salt-ssh)
  Using cached idna-2.6-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests>=1.0.0->salt-ssh)
  Using cached certifi-2017.11.5-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2 (from requests>=1.0.0->salt-ssh)
  Using cached chardet-3.0.4-py2.py3-none-any.whl
Installing collected packages: MarkupSafe, msgpack-python, urllib3, idna, certifi, chardet, requests, tornado, Jinja2, PyYAML, salt-ssh
Successfully installed Jinja2-2.10 MarkupSafe-1.0 PyYAML-3.12 certifi-2017.11.5 chardet-3.0.4 idna-2.6 msgpack-python-0.4.8 requests-2.18.4 salt-ssh-2017.7.2 tornado-4.5.2 urllib3-1.22
(ve) 06:53 kuno@18-volt:~/code/salt$ salt-ssh --version
salt-ssh 2017.7.2 (Nitrogen)
```
5. Created a roster file, but I don't want to put it in /etc/salt/roster.  Instructions say to use --roster-file, but I don't want to specify that every time, so I find out about Saltfile and create this:

```sh
(ve) 07:07 kuno@18-volt:~/code/salt$ cat ./Saltfile
salt-ssh:
  config_dir: "/home/kuno/code/salt/"
  log_file: "/home/kuno/code/salt/ssh.log"
  roster_file: "/home/kuno/code/salt/roster.yaml"
```

6. At https://docs.saltstack.com/en/getstarted/ssh/connect.html now, trying to run test.ping:

```sh
(ve) 07:04 kuno@18-volt:~/code/salt$ salt-ssh -i '*' test.ping
No permissions to access "/var/log/salt/ssh", are you running as the correct user?
```

7. I did specify a log_file in Saltfile, is it even reading that?

```sh
(ve) 07:07 kuno@18-volt:~/code/salt$ salt-ssh --log-level=debug -i '*' test.ping
[INFO    ] Loading Saltfile from '/home/kuno/code/salt/Saltfile'
[DEBUG   ] Reading configuration from /home/kuno/code/salt/Saltfile
[DEBUG   ] Missing configuration file: /home/kuno/code/salt/master
No permissions to access "/var/log/salt/ssh", are you running as the correct user?
```

8. Yes, it is reading the Saltfile, but log_file doesn't work apparently.  Let's try on the commandline:

```sh
(ve) 07:08 kuno@18-volt:~/code/salt$ salt-ssh --log-level=debug --log-file=/home/kuno/code/salt/ssh.log -i '*' test.ping
[INFO    ] Loading Saltfile from '/home/kuno/code/salt/Saltfile'
[DEBUG   ] Reading configuration from /home/kuno/code/salt/Saltfile
[DEBUG   ] Missing configuration file: /home/kuno/code/salt/master
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] LazyLoaded flat.targets
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] compile template: /home/kuno/code/salt/roster.yaml
[DEBUG   ] Jinja search path: ['/var/cache/salt/master/files/base']
[PROFILE ] Time (in seconds) to render '/home/kuno/code/salt/roster.yaml' using 'jinja' renderer: 0.00418400764465332
[DEBUG   ] Rendered data from file: /home/kuno/code/salt/roster.yaml:
jazz:
  host: kuno.frl
  user: kuno

[DEBUG   ] Results of YAML rendering: 
OrderedDict([('jazz', OrderedDict([('host', 'kuno.frl'), ('user', 'kuno')]))])
[PROFILE ] Time (in seconds) to render '/home/kuno/code/salt/roster.yaml' using 'yaml' renderer: 0.0012705326080322266
[DEBUG   ] Matched minions: {'jazz': {'user': 'kuno', 'host': 'kuno.frl'}}
salt-ssh could not be run because it could not generate keys.

You can probably resolve this by executing this script with increased permissions via sudo or by running as root.
You could also use the '-c' option to supply a configuration directory that you have permissions to read and write to.
```

9. Ok, one step further.  The -c option hint seems bogus, as we're already specifying that in the Saltfile.  Let's try anyway because log_file didn't work:

```sh
(ve) 07:30 kuno@18-volt:~/code/salt$ salt-ssh -c /home/kuno/code/salt --log-level=trace --log-file=/home/kuno/code/salt/ssh.log -i '*' test.ping

    (same output as before, -c didn't help)
```
10. Doing some Googling I find that I may need to set pki_dir and cachedir when running as a non-root user (at https://docs.saltstack.com/en/latest/topics/ssh/#running-salt-ssh-as-non-root-user), so let's try that:

```sh
(ve) 07:35 kuno@18-volt:~/code/salt$ mkdir pki
(ve) 07:35 kuno@18-volt:~/code/salt$ mkdir cache
(ve) 07:35 kuno@18-volt:~/code/salt$ cat Saltfile
salt-ssh:
  config_dir: "/home/kuno/code/salt/"
  log_file: "/home/kuno/code/salt/ssh.log"
  pki_dir: "/home/kuno/code/salt/pki"
  cachedir: "/home/kuno/code/salt/cache"
  roster_file: "/home/kuno/code/salt/roster.yaml"
(ve) 07:35 kuno@18-volt:~/code/salt$ salt-ssh --log-level=trace --log-file=/home/kuno/code/salt/ssh.log -i '*' test.ping

    (same output as before, could not generate keys)
```

11. That didn't help, maybe those need to go in the Master file?  The tutorial hasn't mentioned a master file, but I found references to it when trying to solve these issues.  I don't have a master file, so I'll create a minimal one from https://docs.saltstack.com/en/latest/ref/configuration/examples.html:

```sh
(ve) 07:44 kuno@18-volt:~/code/salt$ cat master
root_dir: "/home/kuno/code/salt/"
pki_dir: "pki"
cachedir: "cache"
log_file: "ssh.log"
(ve) 07:44 kuno@18-volt:~/code/salt$ salt-ssh -i '*' test.ping
jazz:
    True
```

12. Hooray!

Probably the main thing which should be clarified is the role of the `master` and `Saltfile` files, and which options go in which file.  It is completely unclear to me why `config_dir` and `roster_file` can go in the Saltfile, but all the other paths need to be in the master file.

**Comment by Ch3LL** [comment]
looks like you might be running into this: https://github.com/saltstack/salt/issues/38336 

There are some current workarounds there as well.

Does this issue cover all the issues you were running into?

**Comment by warpr** [comment]
Hi @Ch3LL , yes it seems #38336 covers most of it.  As long as this:

> Document all acceptable values from parsers.py in a Saltfile documentation.

Also clarifies that other values should go in the master file.  (Or some other way to let readers know about both of those files when going through the "get started" documentation).
